src/main/scala/passes/LoopCheck.scala

```diff
@@ -0,0 +1,174 @@
+package fuselang.passes
+
+import scala.{PartialFunction => PF}
+
+//import fuselang.Utils.RichOption
+
+import fuselang.common._
+import ScopeMap._
+import ScopeMapExt._
+import Syntax._
+import Errors._
+import CompilerError._
+import Checker._
+import EnvHelpers._
+
+object LoopChecker {
+
+  def check(p: Prog) = LCheck.check(p)
+  object States extends Enumeration{
+    type States = Value
+    val USE, DEF, DK = Value
+  }
+  import States._
+
+  // Bounds checker environment doesn't need to track any information. Empty
+  // environment that just runs the commands.
+  private case class LEnv(
+    StateMap: ScopedMapExt[Id, States] = ScopedMapExt(),
+    NameMap: ScopedMap[Id,Id] = ScopedMap())(implicit val res: Int) 
+    extends ScopeManager[LEnv] {
+    
+    
+    // Helper functions for NameMap
+    def addName(vid:Id, tid:Id):LEnv = NameMap.add(vid,tid) match{
+      case None => throw Impossible("NameMap has this view id before, redefinition")
+      case Some(m) => LEnv(StateMap,m)
+    }
+    // in the case of for loop i,  
+    def getName(aid:Id):Id = NameMap.get(aid).getOrElse(aid)
+    
+    // Helper functions for StateMap
+    def atDef(id:Id): LEnv = StateMap.get(id) match{
+      case None | Some(DK) => LEnv(StateMap+(id -> DEF), NameMap)
+      case Some(DEF) => this 
+      case Some(USE) => throw LoopDepSequential(id)
+      case _ => throw Impossible("No such state")
+    }
+    def atDk(id:Id):LEnv = StateMap.get(id) match{
+      case None | Some(DEF) => LEnv(StateMap+(id -> DK), NameMap)
+      case Some(DK) => this
+      case Some(USE) => throw LoopDepSequential(id) 
+      case _ => throw Impossible("No such state")
```

Comment:
Why do you need the `case _`? It seems like you're matching all possible enum cases.

If this is because of exhaustiveness checking warning, you can use `case object`s instead of an enumeration:
```
sealed trait States
case object Def extends States
case object Use extends States
case object DontKnow extends States
```